querybook/server/lib/table_upload/exporter/s3_exporter.py

```diff
@@ -0,0 +1,144 @@
+import tempfile
+from typing import Tuple
+import re
+from abc import abstractmethod
+
+import pandas as pd
+import numpy as np
+
+from app.db import with_session
+from clients.s3_client import S3FileCopier
+from clients.s3_client import MultiPartUploader
+
+from lib.utils.execute_query import execute_query
+from lib.table_upload.common import ImporterResourceType
+from logic.admin import get_query_engine_by_id
+from lib.query_analysis.create_table.create_table import (
+    get_external_create_table_statement,
+)
+from lib.table_upload.exporter.utils import update_pandas_df_column_name_type
+from .base_exporter import BaseTableUploadExporter
+
+S3_OBJECT_KEY_NOT_ALLOWED_CHAR = r"[^\w-]+"
+
+
+class S3BaseExporter(BaseTableUploadExporter):
+    @abstractmethod
+    def UPLOAD_FILE_TYPE(cls) -> str:
+        """Override this to specify what kind of file is getting uploaded
+
+        Returns:
+            str: Example: 'CSV' or 'PARQUET'
+        """
+        return NotImplementedError()
+
+    @abstractmethod
+    def upload_to_s3(self) -> None:
```

Comment:
i think abstract private method would break the rule of inheritance: https://www.tutorialspoint.com/can-we-declare-an-abstract-method-private-protected-public-or-default-in-java#:~:text=If%20a%20method%20of%20a,abstract%20method%20cannot%20be%20private.

Reply:
I think that link is specific to Java. I was suggesting prepending the method name with an underscore to indicate that it is just used within the class (like you did with several of the methods in `BaseCreateTable` - `_get_create_prefix`, etc) - not a huge deal though if you decide not to change it

Reply:
thats a good point, they should be `protected` instead of private. Will add underscore then